xla/service/while_loop_all_reduce_code_motion.cc

```diff
@@ -141,6 +153,175 @@ HloInstruction* GetEffectiveScalar(HloInstruction* instruction) {
+    std::optional<int64_t> idx = GetLoopInductionVarTupleIdx(while_op);
+    if (!idx.has_value() || (indvar_idx.has_value() && *indvar_idx != *idx)) {
+      VLOG(5) << "Unable to identify the loop induction variable.";
+      return std::nullopt;
+    }
+    const HloInstruction* init = while_op->operand(0)->operand(*idx);
+    if (!IsZero(init)) {
+      VLOG(5) << "Loop induction variable is not initialized to zero.";
+      return std::nullopt;
+    }
+    indvar_idx = *idx;
+
+    auto range = MatchLoopRangeWithKnownValues(while_op, *indvar_idx,
+                                               /*known_values=*/{});
+    if (!range.has_value() || !range->IsLinear() || !range->IsBounded() ||
+        !range->IsStepKnown() || range->step()->GetSignedValue() != 1 ||
+        (loop_bound.has_value() &&
+         *loop_bound != range->max()->GetSignedValue())) {
+      VLOG(5) << "Only simple range loops are supported.";
+      return std::nullopt;
+    }
+    loop_bound = range->max()->GetSignedValue();
+  }
+
+  // Verify that the instruction is an iterator over the loop range.
+  // Allow reverse indexing (i.e. back-to-front).
+  auto is_induction_variable = [&](const HloInstruction* instr) {
+    const HloInstruction* left = nullptr;
+    const HloInstruction* right = nullptr;
+    if (Match(instr, m::Subtract(m::ConstantScalar(&left), m::Op(&right))) &&
+        left->literal().GetFirstInteger().value() == *loop_bound) {
+      instr = right;
+    }
+    return Match(instr, m::GetTupleElement(m::Parameter(0), *indvar_idx));
+  };
+
+  // Do not hoist all-reduce ops if the resulting all-reduce is too large.
+  // The threshold value is chosen arbitrarily.
+  const int64_t kAllReduceMaxSizeBytes = 10240;
+  if ((*loop_bound + 1) * ShapeUtil::ArraySize(all_reduce->shape()) >
+      kAllReduceMaxSizeBytes) {
+    VLOG(5) << "Resulting all-reduce exceeds the size threshold: "
+            << kAllReduceMaxSizeBytes;
+    return std::nullopt;
+  }
+
+  // Get user paths of the all-reduce.
+  auto user_paths = gpu::GetSlicedUserPaths(
+      *all_reduce, *call_graph,
+      HloPredicateIsOp<HloOpcode::kReshape, HloOpcode::kConvert>,
+      /*check_alignment=*/false,
+      /*allow_multiple_callers=*/true);
+  if (user_paths.size() != 1) {
+    VLOG(5) << "Only single user of the all-reduce is supported.";
+    return std::nullopt;
+  }
+
+  // Process HLO ops preceding the DUS.
+  auto& user_chain = user_paths.back();
+  const HloInstruction* hlo = user_chain.back();
+  user_chain.pop_back();
+  for (HloInstruction* hlo : user_chain) {
+    switch (hlo->opcode()) {
+      case HloOpcode::kReshape:
+        context.update_slice_shape = hlo->shape();
+        break;
+      case HloOpcode::kConvert:
+        context.update_slice_shape.set_element_type(
+            hlo->shape().element_type());
+        break;
+      default:
+        return std::nullopt;
+    }
+  }
+
+  // DUS ops may form a chain.
+  const HloInstruction* prev =
```

Comment:
Is this true in practice that you can have such a long chain of dynamic update slices? DUS (DUS (DUS ...)))?